sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -388,57 +445,18 @@ class Analyzer(
-        logDebug(s"Conflicting attributes ${conflictingAttributes.mkString(",")} in $j")
-
-        right.collect {
-          // Handle base relations that might appear more than once.
-          case oldVersion: MultiInstanceRelation
-              if oldVersion.outputSet.intersect(conflictingAttributes).nonEmpty =>
-            val newVersion = oldVersion.newInstance()
-            (oldVersion, newVersion)
-
-          // Handle projects that create conflicting aliases.
-          case oldVersion @ Project(projectList, _)
-              if findAliases(projectList).intersect(conflictingAttributes).nonEmpty =>
-            (oldVersion, oldVersion.copy(projectList = newAliases(projectList)))
-
-          case oldVersion @ Aggregate(_, aggregateExpressions, _)
-              if findAliases(aggregateExpressions).intersect(conflictingAttributes).nonEmpty =>
-            (oldVersion, oldVersion.copy(aggregateExpressions = newAliases(aggregateExpressions)))
-
-          case oldVersion: Generate
-              if oldVersion.generatedSet.intersect(conflictingAttributes).nonEmpty =>
-            val newOutput = oldVersion.generatorOutput.map(_.newInstance())
-            (oldVersion, oldVersion.copy(generatorOutput = newOutput))
-
-          case oldVersion @ Window(_, windowExpressions, _, _, child)
-              if AttributeSet(windowExpressions.map(_.toAttribute)).intersect(conflictingAttributes)
-                .nonEmpty =>
-            (oldVersion, oldVersion.copy(windowExpressions = newAliases(windowExpressions)))
-        }
-        // Only handle first case, others will be fixed on the next pass.
-        .headOption match {
-          case None =>
-            /*
-             * No result implies that there is a logical plan node that produces new references
-             * that this rule cannot handle. When that is the case, there must be another rule
-             * that resolves these conflicts. Otherwise, the analysis will fail.
-             */
-            j
-          case Some((oldRelation, newRelation)) =>
-            val attributeRewrites = AttributeMap(oldRelation.output.zip(newRelation.output))
-            val newRight = right transformUp {
-              case r if r == oldRelation => newRelation
-            } transformUp {
-              case other => other transformExpressions {
-                case a: Attribute => attributeRewrites.get(a).getOrElse(a)
-              }
-            }
-            j.copy(right = newRight)
-        }
+      // To resolve duplicate expression IDs for all the BinaryNode
+      case b: BinaryNode if !b.duplicateResolved => b match {
+        case j @ Join(left, right, _, _) =>
+          j.copy(right = dedupRight(left, right))
+        case i @ Intersect(left, right) =>
+          i.copy(right = dedupRight(left, right))
+        case e @ Except(left, right) =>
+          e.copy(right = dedupRight(left, right))
+        case cg: CoGroup =>
```

Comment:
@cloud-fan After code changes, I just realized the problem. If we replace intersect with left-semi join during analysis, it does not work. We have to de-duplicate the left and the right children at first; otherwise, the tree will be like
```
== Analyzed Logical Plan ==
id: int
Distinct
+- Join LeftSemi, Some((id#1 <=> id#1))
   :- Sample 0.0, 0.4, false, 1
   :  +- Sort [id#1 ASC], false
   :     +- Project [_1#0 AS id#1]
   :        +- LogicalRDD [_1#0], MapPartitionsRDD[2] at apply at Transformer.scala:22
   +- Sample 0.4, 1.0, false, 1
      +- Sort [id#4 ASC], false
         +- Project [_1#0 AS id#4]
            +- LogicalRDD [_1#0], MapPartitionsRDD[2] at apply at Transformer.scala:22
```
The Join condition is invalid. Should we keep the rule in the optimizer? 

BTW, I will remove the `Except` and `CoGroup`.